File: лаб4/tree.cpp
#include <vector>
#include <iostream>
using namespace std;

// Узел общего дерева
class TreeNode {
public:
    int value;                      // Значение узла
    vector<TreeNode*> children;     // Вектор дочерних узлов
    
    // Конструктор
    TreeNode(int val) : value(val) {}
    
    // Добавление дочернего узла
    void addChild(TreeNode* child) {
        children.push_back(child);
    }
    
    // Удаление дочернего узла
    void removeChild(TreeNode* child) {
        for (auto it = children.begin(); it != children.end(); ++it) {
            if (*it == child) {
                children.erase(it);
                break;
            }
        }
    }
    
    // Обход дерева в глубину
    void traverse() {
        cout << value << " ";
        for (TreeNode* child : children) {
            child->traverse();
        }
    }
};

// Узел бинарного дерева
class BinaryTreeNode {
public:
    int value;                      // Значение узла
    BinaryTreeNode* left;           // Левый потомок
    BinaryTreeNode* right;          // Правый потомок
    BinaryTreeNode* parent;         // Родительский узел
    
    // Конструктор
    BinaryTreeNode(int val) : value(val), left(nullptr), right(nullptr), parent(nullptr) {}
    
    // Вставка левого потомка
    BinaryTreeNode* insertLeft(int val) {
        left = new BinaryTreeNode(val);
        left->parent = this;
        return left;
    }
    
    // Вставка правого потомка
    BinaryTreeNode* insertRight(int val) {
        right = new BinaryTreeNode(val);
        right->parent = this;
        return right;
    }
    
    // Деструктор (для очистки памяти)
    ~BinaryTreeNode() {
        delete left;
        delete right;
    }
};

File: лаб4/tree.java
import java.util.ArrayList;
import java.util.List;

// Узел общего дерева
class TreeNode {
    public int value;                    // Значение узла
    public List<TreeNode> children;      // Список дочерних узлов
    
    // Конструктор
    public TreeNode(int value) {
        this.value = value;
        this.children = new ArrayList<>();
    }
    
    // Добавление дочернего узла
    public void addChild(TreeNode child) {
        children.add(child);
    }
    
    // Удаление дочернего узла
    public void removeChild(TreeNode child) {
        children.remove(child);
    }
    
    // Обход дерева в глубину
    public void traverse() {
        System.out.print(value + " ");
        for (TreeNode child : children) {
            child.traverse();
        }
    }
}

// Узел бинарного дерева
class BinaryTreeNode {
    public int value;                    // Значение узла
    public BinaryTreeNode left;          // Левый потомок
    public BinaryTreeNode right;         // Правый потомок
    public BinaryTreeNode parent;        // Родительский узел
    
    // Конструктор
    public BinaryTreeNode(int value) {
        this.value = value;
        this.left = null;
        this.right = null;
        this.parent = null;
    }
    
    // Вставка левого потомка
    public BinaryTreeNode insertLeft(int value) {
        left = new BinaryTreeNode(value);
        left.parent = this;
        return left;
    }
    
    // Вставка правого потомка
    public BinaryTreeNode insertRight(int value) {
        right = new BinaryTreeNode(value);
        right.parent = this;
        return right;
    }
    
    // Обход в порядке in-order (для бинарного дерева)
    public void inOrderTraversal() {
        if (left != null) {
            left.inOrderTraversal();
        }
        System.out.print(value + " ");
        if (right != null) {
            right.inOrderTraversal();
        }
    }
}
